sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -16,189 +15,324 @@
-            Duration.ofSeconds(60), 0);
 
     public GoneAndRetryWithRetryPolicy(RxDocumentServiceRequest request, Integer waitTimeInSeconds) {
-        this.request = request;
+        this.goneRetryPolicy = new GoneRetryPolicy(request, waitTimeInSeconds, durationTimer);
+        this.retryWithRetryPolicy = new RetryWithRetryPolicy(request, waitTimeInSeconds, this.durationTimer);
         startStopWatch(this.durationTimer);
-        if (waitTimeInSeconds != null) {
-            this.waitTimeInSeconds = waitTimeInSeconds;
-        } else {
-            this.waitTimeInSeconds = DEFAULT_WAIT_TIME_IN_SECONDS;
-        }
     }
 
     @Override
     public Mono<ShouldRetryResult> shouldRetry(Exception exception) {
-        CosmosException exceptionToThrow = null;
-        Duration backoffTime = Duration.ofSeconds(0);
-        Duration timeout = Duration.ofSeconds(0);
-        boolean forceRefreshAddressCache = false;
-        if (!(exception instanceof GoneException) &&
-            !(exception instanceof RetryWithException) &&
-            !(exception instanceof PartitionIsMigratingException) &&
-            !(exception instanceof InvalidPartitionException &&
-                (this.request.getPartitionKeyRangeIdentity() == null ||
-                this.request.getPartitionKeyRangeIdentity().getCollectionRid() == null)) &&
-            !(exception instanceof PartitionKeyRangeIsSplittingException)) {
-
-            logger.warn("Operation will NOT be retried. Current attempt {}, Exception: ", this.attemptCount,
-                exception);
-            stopStopWatch(this.durationTimer);
-            return Mono.just(ShouldRetryResult.noRetry());
-        } else if (exception instanceof GoneException &&
-            !request.isReadOnly() &&
-            BridgeInternal.hasSendingRequestStarted((CosmosException)exception)) {
+
+        return this.retryWithRetryPolicy.shouldRetry(exception)
+                                        .flatMap((retryWithResult) -> {
+
+            if (retryWithResult.shouldRetry) {
+                return Mono.just(retryWithResult);
+            }
+
+            return this.goneRetryPolicy.shouldRetry(exception)
+                .flatMap((goneRetryResult) -> {
+                    if (!goneRetryResult.shouldRetry) {
+                        logger.warn("Operation will NOT be retried. Exception:",
+                            exception);
+                        stopStopWatch(this.durationTimer);
+                    }
+
+                    return Mono.just(goneRetryResult);
+                });
+        });
+    }
+
+    private void stopStopWatch(StopWatch stopwatch) {
+        synchronized (stopwatch) {
+            stopwatch.stop();
+        }
+    }
+
+    private void startStopWatch(StopWatch stopwatch) {
+        synchronized (stopwatch) {
+            stopwatch.start();
+        }
+    }
+
+    static class GoneRetryPolicy extends RetryPolicyWithDiagnostics {
```

Comment:
NON-BLOCKER: Is exponential retry used in many retry policies, if so would it help if its aslo refractored out?

Reply:
Agree that this should be considered in the future